controllers/edgedeployment_controller_test.go

```diff
@@ -0,0 +1,509 @@
+			EdgeDeploymentRepository: deployRepoMock,
+			EdgeDeviceRepository:     edgeDeviceRepoMock,
+		}
+
+		signalContext, cancelContext = context.WithCancel(context.TODO())
+		go func() {
+			err = k8sManager.Start(signalContext)
+			Expect(err).ToNot(HaveOccurred())
+		}()
+
+		req = ctrl.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      "test",
+				Namespace: "test",
+			},
+		}
+	})
+
+	AfterEach(func() {
+		cancelContext()
+		mockCtrl.Finish()
+	})
+
+	getDevice := func(name string) *v1alpha1.EdgeDevice {
+		return &v1alpha1.EdgeDevice{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      name,
+				Namespace: "test",
+			},
+			Spec: v1alpha1.EdgeDeviceSpec{
+				OsImageId:   "test",
+				RequestTime: &v1.Time{},
+				Heartbeat:   &v1alpha1.HeartbeatConfiguration{},
+			},
+		}
+	}
+
+	Context("Reconcile", func() {
+		It("Return nil if no edgedeployment found", func() {
+			// given
+			return_erro := errors.NewNotFound(schema.GroupResource{Group: "", Resource: "notfound"}, "notfound")
+			deployRepoMock.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, return_erro).AnyTimes()
+			// when
+			res, err := edgeDeploymentReconciler.Reconcile(context.TODO(), req)
+
+			// then
+			Expect(err).NotTo(HaveOccurred())
+			Expect(res).To(Equal(reconcile.Result{Requeue: false, RequeueAfter: 0}))
+		})
+
+		It("Return error if edgedeployment retrieval failed", func() {
+			// given
+			deployRepoMock.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, fmt.Errorf("Failed")).AnyTimes()
+
+			// when
+			res, err := edgeDeploymentReconciler.Reconcile(context.TODO(), req)
+
+			// then
+			Expect(err).To(HaveOccurred())
+			Expect(res).To(Equal(reconcile.Result{Requeue: true, RequeueAfter: 0}))
+		})
+
+		Context("Finalizers", func() {
+			var (
+				deploymentData *v1alpha1.EdgeDeployment
+			)
+
+			BeforeEach(func() {
+				deploymentData = &v1alpha1.EdgeDeployment{Spec: v1alpha1.EdgeDeploymentSpec{
+					DeviceSelector: &v1.LabelSelector{
+						MatchLabels: map[string]string{"test": "test"},
+					},
+					Device: "test",
+					Type:   "test",
+					Pod:    v1alpha1.Pod{},
+					Data:   &v1alpha1.DataConfiguration{},
+				}}
+
+				deployRepoMock.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any()).
+					Return(deploymentData, nil).Times(1)
+
+			})
+
+			It("Added finalizer requeue correctly", func() {
+				// given
+				deployRepoMock.EXPECT().Patch(gomock.Any(), deploymentData, gomock.Any()).
+					Return(nil).Do(func(ctx context.Context, old, new *v1alpha1.EdgeDeployment) {
+					Expect(new.Finalizers).To(HaveLen(1))
+					Expect(old.Finalizers).To(HaveLen(0))
+				}).Times(1)
+
+				// when
+				res, err := edgeDeploymentReconciler.Reconcile(context.TODO(), req)
+
+				// then
+				Expect(err).NotTo(HaveOccurred())
+				Expect(res).To(Equal(reconcile.Result{Requeue: true, RequeueAfter: 0}))
+			})
+
+			It("Added finalizer failed requeue with error", func() {
```

Comment:
What do you think about adding delete device and see how finalizers are removed? Can be done in separate PR